### PR TITLE
[DDP] Support not all outputs used in loss calculation

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57081 [DDP] Support not all outputs used in loss calculation**

Changes in this diff:
1. Enable passthrough autograd function when find_unused_parameters=True.
2. With above, move `prepare_for_backward` which does unused parameter checking logic to beginning of backwards pass, only when `find_unused_parameters=True`. 
3. Enhance process of unused parameter checking to account for outputs not being used in loss.

The way (3) is implemented is by triggering the autograd hook corresponding to parameters that did not participate in loss computation. Since they did not participate, the autograd hook is triggered with a gradient of None, and the reducer handles this appropriately to ensure that the gradient is not touched.

Tested by ensuring that when a model output is not used in loss, the corresponding grad is not modified. Also verified that the grads are the same in local vs DDP training case. Also verified that gradients are not touched in this case, i.e. if grad is originally `None`, it stays as `None`, not zero, after.

Note that in this diff we are not enabling the pass through autograd function for regular case find_unused_parameters=False because that has a much bigger blast radius and needs additional careful analysis especially with regard to the performance. 

Differential Revision: [D28048628](https://our.internmc.facebook.com/intern/diff/D28048628/)

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -800,33 +835,32 @@ def forward(self, *inputs, **kwargs):
             else:
                 output = self.module(*inputs, **kwargs)
 
+            disable = False
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:838]: why is this flag needed?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -800,33 +835,32 @@ def forward(self, *inputs, **kwargs):
...
             else:
                 self.require_forward_param_sync = False
 
-        # TODO. Right now we add this sink for static_graph training only. once
-        # this feature is stable, we will add this sink for all cases. E.g.
-        # This sink can help capture more accuracte backward start time as well.
-        if self.static_graph and self.num_iterations == 1:
-            # Need to grab list of tensors from user output in order to pass
-            # to custom autograd function.
+        if not disable and (self.find_unused_parameters and not self.static_graph) or (self.static_graph and self.num_iterations == 1):
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:849]: line too long

File: torch/csrc/distributed/c10d/init.cpp [hunk 3]
```diff
@@ -401,6 +401,10 @@ An enum-like class for built-in communication hooks: ``ALLREDUCE`` and ``FP16_CO
           "save_thread_local_state",
           &::c10d::Reducer::save_thread_local_state,
           py::call_guard<py::gil_scoped_release>())
+      .def(
+          "set_per_iteration_param_outputs_unused",
```
Comment (zhaojuanmao) [inline torch/csrc/distributed/c10d/init.cpp:405]: nit: make it private

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -800,33 +835,32 @@ def forward(self, *inputs, **kwargs):
...
-                # this forward pass, to ensure we short circuit reduction for any
-                # unused parameters. Only if `find_unused_parameters` is set.
-                if self.find_unused_parameters and not self.static_graph:
-                    # Do not need to populate this for static graph.
-                    self.reducer.prepare_for_backward(list(_find_tensors(output)))
-                else:
-                    self.reducer.prepare_for_backward([])
+                if disable or not self.find_unused_parameters or self.static_graph:
+                    if self.find_unused_parameters:
+                        self.reducer.prepare_for_backward(list(_find_tensors(output)))
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:843]: in which case this will happen?

File: torch/nn/parallel/distributed.py [hunk 5]
```diff
@@ -800,33 +835,32 @@ def forward(self, *inputs, **kwargs):
...
+                    else:
+                        self.reducer.prepare_for_backward([])
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:845]: should it be just like this?

if self.static_graph or not self.find_unused_parameters:
    self.reducer.prepare_for_backward([])

Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:845]: Yep - this is much clearer. Thanks!

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:838]: Thanks for flagging, accidentally send debug code. Removing.

File: torch/nn/parallel/distributed.py (hunk 4 above)
Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:843]: Yeah, with this change we should never need to call `self.reducer.prepare_for_backward(list(_find_tensors(output)))`. The suggestion below is much clearer.

File: torch/nn/parallel/distributed.py [hunk 6]
```diff
@@ -113,14 +113,31 @@ class _DDPUnevenInputsConfig(NamedTuple):
...
+        elif state_dict['find_unused'] and not static_graph_training:
+            # First type of unused params: parameters that did not participate
+            # in computing model outputs. These are found by the below call to
+            # prepare_for_backward.
+            # Second type of unused params: params that won't get gradient
+            # because outputs they produced do not get used in computing loss
+            # for this call to backward. Due to this passthrough autograd
+            # function, autograd hooks for these parameters are now triggered
+            # with zero gradient to maintain parity with local training.
+            ctx.reducer.prepare_for_backward(list(_find_tensors(ctx.inputs)))
```
Comment (zhaojuanmao) [inline torch/nn/parallel/distributed.py:138]: does it mean the unused parameters due to unused output will have zero gradients instead of being 'undefined'? I'm not sure whether this is OK?
also I'm curious about why gradients of unused output tensors are not None..?

Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:138]: @zhaojuanmao Yes, if a parameter goes unused because it wasn't used as part of loss computation it will have a zero gradient. This is what the autograd custom function does even on a local model, if it does not receive a grad for the tensor in the backwards pass, it fills it in with zero, see: 

Although, IIUC this shouldn't make a difference? If the grad for the param is k, then we accumulate a grad of `0` onto it and the grad is the same. On the other hand, if grad is `None`, then it will be changed to `0`, but this would happen with `model.zero_grad()` anyways?

Comment (rohan-varma) [reply torch/nn/parallel/distributed.py:143]: @zhaojuanmao This should be fixed now with the above ` ctx.set_materialize_grads(False)`.

File: torch/lib/c10d/reducer.cpp [hunk 7]
```diff
@@ -526,7 +536,18 @@ void Reducer::autograd_hook(VariableIndex index) {
     // to mark it in local_used_maps_. During no_sync session, the same var can
     // be set multiple times, which is OK as does not affect correctness. As
     // long as it is used once during no_sync session, it is marked as used.
-    local_used_maps_[index.replica_index][index.variable_index] = 1;
+    // Only set it as locally used if the grad is defined. Otherwise, hooks
+    // could sometimes be triggered with undefined grads, and if this happens
+    // globally, we don't want to touch the .grad field of the param.
+
+    auto& variable = get_param_from_index(index);
+    runGradCallbackForVariable(variable, [&](auto& grad) {
```
Comment (zhaojuanmao) [inline torch/lib/c10d/reducer.cpp:544]: why the runGradCallbackForVariable is needed here? grad is not mutated anyway

**Review by zhaojuanmao — APPROVED** [review]
lgtm, thanks!

File: torch/lib/c10d/reducer.cpp (hunk 7 above)
Comment (rohan-varma) [reply torch/lib/c10d/reducer.cpp:544]: We have to access `variable.grad()` right? When running in just DDP, this can be done with `variable.grad()` directly, but I believe this callback is needed when running in RPC context, where the grad is not available in variable.grad() field.

Comment (zhaojuanmao) [reply torch/lib/c10d/reducer.cpp:544]: a, you are right!
